tests/10.ABI.ts

```diff
@@ -218,214 +219,228 @@ describe('ABI type checking', () => {
+        3,
+      ])
+    ),
+    newTestCase(
+      new ABIArrayDynamicType(new ABIBoolType()),
+      [],
+      new Uint8Array([0, 0])
+    ),
+    newTestCase(
+      new ABIArrayDynamicType(new ABIBoolType()),
+      [true, true, false],
+      new Uint8Array([0, 3, 192])
+    ),
+    newTestCase(
+      new ABIArrayDynamicType(new ABIBoolType()),
+      [false, true, false, false, false, false, false, false],
+      new Uint8Array([0, 8, 64])
+    ),
+    newTestCase(
+      new ABIArrayDynamicType(new ABIBoolType()),
+      [true, false, false, true, false, false, true, false, true],
+      new Uint8Array([0, 9, 146, 128])
+    ),
+    newTestCase(ABIType.from('()'), [], new Uint8Array([])),
+    // 2^6 + 2^5 = 64 + 32 = 96
+    newTestCase(
+      ABIType.from('(bool,bool,bool)'),
+      [false, true, true],
+      new Uint8Array([96])
+    ),
+    newTestCase(
+      ABIType.from('(bool[3])'),
+      [[false, true, true]],
+      new Uint8Array([96])
+    ),
+    newTestCase(
+      ABIType.from('(bool[])'),
+      [[false, true, true]],
+      new Uint8Array([0, 2, 0, 3, 96])
+    ),
+    newTestCase(
+      ABIType.from('(bool[2],bool[])'),
+      [
+        [true, true],
+        [true, true],
+      ],
+      new Uint8Array([192, 0, 3, 0, 2, 192])
+    ),
+    newTestCase(
+      ABIType.from('(bool[],bool[])'),
+      [[], []],
+      new Uint8Array([0, 4, 0, 6, 0, 0, 0, 0])
+    ),
+    newTestCase(
+      ABIType.from('(string,bool,bool,bool,bool,string)'),
+      ['AB', true, false, true, false, 'DE'],
+      new Uint8Array([0, 5, 160, 0, 9, 0, 2, 65, 66, 0, 2, 68, 69])
+    ),
+  ].forEach((testCase) => {
+    it(`should round-trip ${testCase.abiType}, ${testCase.input}`, () => {
+      const encoded = testCase.abiType.encode(testCase.input);
+      assert.deepStrictEqual(encoded, testCase.expectedEncoding);
+      const decoded = testCase.abiType.decode(encoded);
+
+      // Converts any numeric type to BigInt for strict equality comparisons.
+      // The conversion is required because there's no type information
+      // available to convert a decoded BigInt back to its original number
+      // form.  Converting from number to BigInt is always _safe_.
+      function numericAsBigInt(d: ABIValue): ABIValue {
+        if (typeof d === 'number') {
+          return BigInt(d);
+        }
+        if (d instanceof Array) {
+          return (d as ABIValue[]).map(numericAsBigInt);
+        }
+        return d;
+      }
+
+      // Returns true when the provided ABIType decodes to BigInt.
+      function decodeReturnsBigInt(t: ABIType): boolean {
+        if (t instanceof ABIUintType || t instanceof ABIUfixedType) {
+          return true;
+        }
+        if (t instanceof ABITupleType) {
+          (t as ABITupleType).childTypes
+            .map(decodeReturnsBigInt)
+            .includes(true);
+        } else if (t instanceof ABIArrayStaticType) {
+          return decodeReturnsBigInt((t as ABIArrayStaticType).childType);
```

Comment:
Just to check: is the `as` necessary here if we check the `instanceof` before it?

Reply:
@algochoi Thanks - You're correct.  The cast is not needed.  I simply didn't realize it, appreciate the pointer.  https://github.com/algorand/js-algorand-sdk/pull/701/commits/78f3e67243520c33581a35da365e72a71fbc2864 attempts to remedy the situation.